### PR TITLE
GH-3032: RecoveryCallback on RetryingDeserializer

### PR DESCRIPTION
**Pull request by sobychacko**
Fixes: #3032

### CODE REVIEW

File: spring-kafka/src/main/java/org/springframework/kafka/support/serializer/RetryingDeserializer.java [hunk 1]
```diff
@@ -57,22 +60,31 @@ public void configure(Map<String, ?> configs, boolean isKey) {
...
 	public void close() {
 		this.delegate.close();
 	}
 
+	/**
+	 * Set a recovery callback to execute when the retries are exhausted.
+	 * @param recoveryCallback {@link RecoveryCallback} to execute
+	 * @since 3.1.2
+	 */
+	public void setRecoveryCallback(@Nullable RecoveryCallback<T> recoveryCallback) {
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/support/serializer/RetryingDeserializer.java:86]: Setters must follow just after constructors.

File: spring-kafka/src/test/java/org/springframework/kafka/support/serializer/RetryingDeserializerTests.java [hunk 2]
```diff
@@ -54,6 +55,39 @@ void testRetry() {
 		rdes.close();
 	}
 
+	@Test
+	void retryDeserializerWithRecoveryCallback() throws InterruptedException {
+		NonRecoverableDeser delegate = new NonRecoverableDeser();
+		RetryingDeserializer<String> rdes = new RetryingDeserializer<>(delegate, new RetryTemplate());
```
Comment (artembilan) [inline spring-kafka/src/test/java/org/springframework/kafka/support/serializer/RetryingDeserializerTests.java:61]: You also can just inject a lambda which would throw that exception.
I don't think it matters how many time we retry.
Only what is important for this test that `RecoveryCallback` is called.

File: spring-kafka/src/test/java/org/springframework/kafka/support/serializer/RetryingDeserializerTests.java [hunk 3]
```diff
@@ -54,6 +55,39 @@ void testRetry() {
...
+		rdes.setRecoveryCallback(context -> {
+			delegate.latch.countDown();
```
Comment (artembilan) [inline spring-kafka/src/test/java/org/springframework/kafka/support/serializer/RetryingDeserializerTests.java:63]: You can just sent here a result of `mock()` and then verify in the end that it was called.
There is no any race conditions: everything in the test is called sequentially.

File: spring-kafka/src/test/java/org/springframework/kafka/support/serializer/RetryingDeserializerTests.java [hunk 4]
```diff
@@ -54,6 +55,39 @@ void testRetry() {
...
+		}
+
+		@Override
+		public String deserialize(String topic, byte[] data) {
+			latch.countDown();
+			throw new RuntimeException();
+		}
+
+		@Override
+		public void close() {
```
Comment (artembilan) [inline spring-kafka/src/test/java/org/springframework/kafka/support/serializer/RetryingDeserializerTests.java:86]: The `Deserializer` is functional interface, so no need to override those `default` methods.
See my comment above about its simple lambda for the `RetryingDeserializer` ctor.

**Review by artembilan — CHANGES_REQUESTED** [review]
Also missed an entry in the `whats-new.adoc`.

Thanks

**Comment by sobychacko** [comment]
> Also missed an entry in the `whats-new.adoc`.
> 
> Thanks

Aren't we going to backport the change to `3.1.x` though?

File: spring-kafka/src/test/java/org/springframework/kafka/support/serializer/RetryingDeserializerTests.java [hunk 5]
```diff
@@ -54,6 +59,19 @@ void testRetry() {
 		rdes.close();
 	}
 
+	@SuppressWarnings("unchecked")
+	@Test
+	void retryingDeserializerWithRecoveryCallback() throws Exception {
+		RetryingDeserializer<String> rdes = new RetryingDeserializer<>((s, b) -> {
+			throw new RuntimeException();
+		}, new RetryTemplate());
+		RecoveryCallback<String> recoveryCallback = mock(RecoveryCallback.class);
```
Comment (artembilan) [inline spring-kafka/src/test/java/org/springframework/kafka/support/serializer/RetryingDeserializerTests.java:68]: I know that is going to be nit-picking, but I'll shoot it anyway.

The Mockito starting with some version does not need the type argument.
Just `RecoveryCallback<String> recoveryCallback = mock();` is enough to make it working.
And the code looks nicer.

File: spring-kafka/src/test/java/org/springframework/kafka/support/serializer/RetryingDeserializerTests.java [hunk 6]
```diff
@@ -54,6 +59,19 @@ void testRetry() {
...
+		rdes.setRecoveryCallback(recoveryCallback);
+		rdes.deserialize("my-topic", "my-data".getBytes());
+		verify(recoveryCallback, times(1)).recover(any(RetryContext.class));
```
Comment (artembilan) [inline spring-kafka/src/test/java/org/springframework/kafka/support/serializer/RetryingDeserializerTests.java:71]: The ` times(1)` is there by default:
```
    public static <T> T verify(T mock) {
        return MOCKITO_CORE.verify(mock, times(1));
    }
```
Why do you use this explicitly?

File: spring-kafka/src/test/java/org/springframework/kafka/support/serializer/RetryingDeserializerTests.java [hunk 7]
```diff
@@ -54,6 +59,19 @@ void testRetry() {
...
+		rdes.close();
```
Comment (artembilan) [inline spring-kafka/src/test/java/org/springframework/kafka/support/serializer/RetryingDeserializerTests.java:72]: There is nothing to close. See our delegate lambda.
IMHO the test should be as short as possible and be focused on the specific task.
If tests are too complicated and they are flaky, it is hard to find the reason and solution.
I know: this one is not, but better to have a good habit for less code - better support!

**Review by artembilan — CHANGES_REQUESTED** [review]

File: spring-kafka/src/test/java/org/springframework/kafka/support/serializer/RetryingDeserializerTests.java (hunk 7 above)
Comment (sobychacko) [reply spring-kafka/src/test/java/org/springframework/kafka/support/serializer/RetryingDeserializerTests.java:72]: ok, will cleanup, i was just following original test patterns.

**Review by artembilan — APPROVED** [review]
LGTM.
Will merger when PR build is green.

**Comment by vspiliop** [comment]
@artembilan How can I find when `spring-kafka v3.1.2` will be included in a 

```
  <parent>
    <groupId>org.springframework.boot</groupId>
    <artifactId>spring-boot-dependencies</artifactId>
    <version>...</version>
  </parent>
```

Should I assume in the next spring boot release? Is there a link with planned future spring boot releases?

Thanks again!

**Pull request merged**